Allow data archiving using the :truncate_data task

sauloperez opened the pull request:
#### What? Why?

Closes #4782 (I hope)

This refactors the `:truncate_data` rake task to allow for the removal of old data from an OFN instance as a mean for data archiving. Older data can still be recovered from an older backup. We just need to know when the data archiving was performed and go find the backup we made right
before.

It extracts the `TruncateData` service to make it possible to test without the annoyance of rake task testing. It's not a straightforward thing as you can see in https://thoughtbot.com/blog/test-rake-tasks-like-a-boss. It's a headache we can avoid if we conform to the testing pyramid.

This also gives us room to iterate on our data archiving strategy as we move along.

All credit goes to @lin-d-hop for insisting on the idea of data archiving.

##### Heads-up
:warning: to truncate data in staging as we used to you'll need to run 

```
$ bundle exec rake "ofn:data:truncate[3]"
```
That's because we switched the default 3 months to 24.

#### What should we test?

Test that loading production data in staging and then truncating it keeps working as expected. There should be no data integrity issues.

#### Release notes

Refactored truncate data rake task and default to 24 months instead of 3.
Changelog Category: Changed

luisramos0 commented:
nice, the first commit is doing too many things. Would you be able to hack the commit history and start with a simple commit that just moves the code from the task to the service? Otherwise it is very painful to see what has actually changed.

sauloperez commented:
I'll do my best to split it into two commits but if take a closer look at the deletion code I don't change a single line. I just work on the `#date` method to allow specifying a date from the command line.

sauloperez commented:
Done. I hope this is now easier to follow @luisramos0

sauloperez commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -1,81 +1,28 @@
+      \n
+      <%= color('This will permanently change DB contents. Please, make a backup first.', :yellow) %>
+      Are you sure you want to proceed? (y/N)
+      MSG
```
This now looks like

![Screenshot from 2020-03-06 17-56-57](https://user-images.githubusercontent.com/762088/76104566-e56bb200-5fd3-11ea-9bbc-c2ae43e8c873.png)

luisramos0 replied:
Is there a color like :blinking_and_large_red instead of :yellow :-)

sauloperez replied:
bold red will do the trick :trollface:

sauloperez replied:
LOL https://github.com/JEG2/highline/blob/bf489b632afd92c68df0b2c1f348afb66b5495aa/test/test_highline.rb#L475

luisramos0 replied:
ahah, now we have to use it :-)

luisramos0 commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -9,10 +9,11 @@ require 'tasks/data/truncate_data'
 
-      TruncateData.new.call
+      months_to_keep = args.months_to_keep.to_i
+      TruncateData.new(months_to_keep: months_to_keep).call
```
can be just       TruncateData.new(months_to_keep).call

sauloperez replied:
Yes, things evolved since I introduced the keyboard argument and now I see this is just redundant :+1:

sauloperez commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -1,81 +1,44 @@
+#
+# Example:
+#
+# $ bundle exec rake "ofn:data:truncate[12]"
```
I'm thinking it's safer to make the 2-years window the default instead of the current 3 months. Thoughts?

luisramos0 replied:
:+1:

luisramos0 replied:
2 years would be 24 here, right?

sauloperez replied:
Oh shit. I need to rest, yes :joy:

luisramos0 commented on lib/tasks/data/truncate_data.rb:
```diff
@@ -0,0 +1,95 @@
+    Spree::ReturnAuthorization.delete_all
+    Spree::StateChange.delete_all
+    Spree::LogEntry.delete_all
+    sql_delete_from "sessions"
```
I am not sure we can just wipe this out for production.

Spree::ReturnAuthorization
this is not something we can do it prod, although the last time I checked with had under 10 records across all DBs.
There must be a reason for this to be here... I wonder if this is necessary for the staging process, otherwise we could remove this one from the task.

The othres, sessions, StateChange and LogEntry we should probably add a date filter to the delete statement.
,

sauloperez replied:
Let me go through what think of each and see if we find a sweet spot:

* No idea what ReturnAuthorizations is. Something related to returned shipments? we need to investigate.
* StateChange: since this is useful for troubleshooting a month data seems enough.
* LogEntry: the same
* Sessions: this affects users that actively purchasing so I think 2 weeks data is more than enough. Others can afford to log in again.

sauloperez replied:
I'm going to go with this.

luisramos0 replied:
there are only 5 spree_return_authorizations in aus live but we cant just delete them, there is one in 2020.
My code is a lazy shortcut, the ret auth are linked to orders.
So all you need to do is:
      sql_delete_from "spree_turn_authorizations" #{where_order_id_in_orders_to_delete}"

I havent tested but I think this will do it :+1:

sauloperez replied:
I just check what these authorizations are. Yep, your suggestion makes sense.

luisramos0 commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -1,81 +1,46 @@
-        exchange_fees where exchange_id
-        in (select id from exchanges #{where_oc_id_in_ocs_to_delete})"
-      sql_delete_from "exchanges #{where_oc_id_in_ocs_to_delete}"
+      exit unless HighLine.new.agree(message) { |q| q.default = "N" }
```
q is a one letter variable

luisramos0 commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -6,6 +6,22 @@ require 'tasks/data/truncate_data'
+# hubs to do their tax declaration.
+#
+# It's a must to perform a backup right before executing this. Then, to allow
+# for a later data recovery we need to keep track of the exact moment this rake
```
I wonder if a timestamp is actually written to the logs when executing the task. Maybe we can log it when initializing the TruncateData service?

sauloperez replied:
that would improve the situation a bit but the way to keep track of the event IMO is through a Datadog event like we do for deploys and provisions, so it doesn't get lost. Unfortunately, AU is not in Datadog, the 2nd best option we have is sending an event to Slack, I think.

luisramos0 replied:
datadog, awesome :+1:

sauloperez commented on lib/tasks/data/truncate_data.rake:
```diff
@@ -1,81 +1,46 @@
+      message = <<-MSG.strip_heredoc
+      \n
+      <% highlighted_message = "This will permanently change DB contents. Please, make a backup first." %>
+      <%= color(highlighted_message, :blink, :on_red) %>
```
This now looks like

![Screenshot from 2020-03-16 18-25-14](https://user-images.githubusercontent.com/762088/76784289-a5ff4b80-67b3-11ea-9920-c46664f06a78.png)

luisramos0 replied:
:-)

sauloperez commented on lib/tasks/data/truncate_data.rb:
```diff
@@ -0,0 +1,109 @@
+    Rails.logger.info("TruncateData started with truncation date #{date}")
+    yield
+    Rails.logger.info("TruncateData finished")
+  end
```
we now see 

```sh
I, [2020-03-16 19:02:34#4746]  INFO -- : TruncateData started with truncation date 2018-03-16 20:02:34 +0100                                                                            
(...)
I, [2020-03-16 19:02:35#4746]  INFO -- : TruncateData finished

sauloperez commented on lib/tasks/data/truncate_data.rb:
```diff
@@ -36,7 +42,7 @@ def remove_transient_data
     Spree::StateChange.delete_all("created_at < '#{1.month.ago.to_date}'")
     Spree::LogEntry.delete_all("created_at < '#{1.month.ago.to_date}'")
-    sql_delete_from "sessions"
+    Session.delete_all("created_at < '#{2.weeks.ago.to_date}'")
```
I tried but this doesn't work with `delete_all` and making it work would require a combination of `where` + `delete_all` which in turn will make it harder to stub things in the test. I prefer to wait for our meeting.

luisramos0 commented:
Hello guys, I wont be able to attend the meeting tomorrow. I leave my thoughts here:

I understand you concerns Maikel :+1:

I am 100% sure you can delete the data from the prod data as long as you keep a backup. You dont need to keep the records in your live database. I think the first point is not valid here.

Re second point, I understand the concern for sure!
- This will only touch 2 year old data, any problems arising will be related to that data. It's very unlikely this data is actually used.
- This process is already very slow, if you do it with AR I think it will never end... we would have to start from scratch with AR which would be too much work I believe.
- This process has already been tested in staging servers with live data lately, all the consistency problems I have seen there were related to data anonymization, none were related to truncation.
- I think we can run this process with live data and put it in a staging server. From there we will be able to see if data consistency is a problem.
- Lastly, in the unlikely event of seeing consistency problems in prod, we will have the backups to mitigate the situation.

sauloperez commented:
Thanks very much @luisramos0 ! We'll go through your ideas :ok_hand:

sauloperez commented:
@mkllnk and I had a chat and agreed on:

* Test this further with production data in my local env
* We'll thoroughly check the impact of data archiving with different thresholds, not only 2 years.
* We'll go through the list of destroy callbacks that may apply to these models and see if they got covered.
* Meanwhile, Aussies will talk to hubs and producers to see who checks old data regularly
* They let them know data will still be there but not online
* Then, we'll go with regular testing.

sauloperez commented:
Putting this on-hold until https://github.com/openfoodfoundation/openfoodnetwork/pull/5022, https://github.com/openfoodfoundation/openfoodnetwork/pull/5050, https://github.com/openfoodfoundation/openfoodnetwork/pull/5030 are out to the masses with release v2.8.1

sauloperez commented:
After #5030 got deployed on April 1st we see a noticeable improvement on the DB usage but there are still nasty outliers

![Screenshot from 2020-04-15 19-25-56](https://user-images.githubusercontent.com/762088/79367966-f7cff880-7f4e-11ea-92f9-c9d4832f92ac.png)
a month data on orders_and_fulfillments report. Note we enabled APM for UK and AU on March 28th, hence the latency increase since then.

I'll wait for our meeting regarding infrastructure & performance to decide what we do with this.

sauloperez commented:
it's been a while since we introduced those performance improvements that blocked this PR. This wasn't discussed in last Friday's performance but I see it could be something to try it out quite easily if need to iterate once on this round of performance improvements. Triumvirate's DBs are rather bloated.

Aside from that, IMO it's a must be take `remove_transient_data` out of this PR and ship separately. These tables are very large and surely affect the overall DB performance for no reason.

![Screenshot from 2020-04-28 09-28-20](https://user-images.githubusercontent.com/762088/80459480-b8cc7a80-8932-11ea-8ddf-f3f7e723ed01.png)
![Screenshot from 2020-04-28 09-28-07](https://user-images.githubusercontent.com/762088/80459486-b9651100-8932-11ea-973c-778f9d3055b9.png)
![Screenshot from 2020-04-28 09-27-54](https://user-images.githubusercontent.com/762088/80459488-b9fda780-8932-11ea-8a2f-59a8b259dde1.png)

Then, we could a bit of dev testing to the rest and merge it, considering it just an improvement over this staging-only rake task, and nothing meant to be run in production.

Matt-Yorkley commented:
I was looking at those tables the other day, they really need a clean-up. 

spree_state_changes seems to make a new record every time an order, payment, or anything with state machine changes, and records the old value and update time. Do we even need it at all? It seems like 8 million junk records.

The sessions table has really ancient records, can we add a simple rake task for deleting sesdions older than 3 months and just run it once a month (via schedule.rb)?

luisramos0 commented:
fyi there's #3608 for that.

sauloperez commented:
Alright. I added one last commit and rebased onto master. Nothing has changed other than the goal. Instead of running this in production, I consider it a refactor it'll come in handy in the future, so we can merge after a bit of dev testing.

This gets us one step closer to archiving old data in production but misses the thorough testing. We can do that if the need comes up again.

luisramos0 commented:
ok, I am ok with that approach :+1:
We should still test it with some prod db locally to detect major issues, I can try this with one of the live DBs.
I will let you know.

luisramos0 commented:
I tested this locally with uk live data.
I loaded the data, switched payment methods to development, truncated data (no errors), verified some tables were (a lot) smaller, run the server, visited shop, checked out an order and went to the backoffice to open that order, all good :+1:
Merging.
